Implemented fromStorage and fromCert client certificates

BlackGad opened the pull request:
## Feature implementation

Implements: https://github.com/NuGet/Home/issues/5773

## Implementation

Details: Implemented ability to add client x509 certificates via NuGet Cli processing.

1) Internally certificates extracted from configuration items on base Command.Execute and stored as singleton inside static ClientCertificates class with ClientCertificates.Store method.
2) On any HttpHandlerResourceV3 creation internal HttpClientHandler filled with configured certificates with ClientCertificates.SetupClientHandler method.
3)Added new nuget.config configuration section **clientCertificates** which may have children of 2 types:

**fromStorage** - item for certificate import from [certificate store](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores). Internally uses [X509Certificate2Collection.Find](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.find?view=netframework-4.8#System_Security_Cryptography_X509Certificates_X509Certificate2Collection_Find_System_Security_Cryptography_X509Certificates_X509FindType_System_Object_System_Boolean_) method.
Can be configured with 4 Add item's

- Optional Key="**StoreLocation**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storelocation?view=netframework-4.8#fields)".  
Equals '**CurrentUser**' by default
- Optional Key="**StoreName**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=netframework-4.8#fields)". 
Equals '**My**' by default
- Optional Key="**FindType**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.8#fields)". 
Equals '**FindByThumbprint**' by default
- Optional Key="**FindValue**" Value="The search criteria as an object"

- **fromFile** - item for certificate import directly from file (DER encoded x.509, Base-64 encoded x.509, PKCS)
Can be configured with 2 Add item's

- Key="**Path**" Value="Absolute or relative path to certificate file". Relative path resolve have 2 stages: first relative to configuration file origin, second relative to current application directory.
- Optional Key="**Password**" Value="Encrypted password". Password for certificate file. Encrypted in same manner as PackageSourceCredential password.

- **fromPEM** - item for certificate import directly from configuration body (Base-64 encoded x.509 in PEM format)
Can be configured with 1 optional Add item

- Items body must  be filled with [PEM](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail) (Base-64 encoded x.509 certificate). Start and end of text are trimmed from spaces and new line chars.

- Optional Key="**Password**" Value="Encrypted password". Password for certificate file. Encrypted in same manner as PackageSourceCredential password.

## Configuration example

```xml
<configuration>
...
    <clientCertificates>	
        <fromStorage name="Some key from storage">
            <!-- Optional. CurrentUser by default -->
            <add key="StoreLocation" value="CurrentUser" />
            <!-- Optional. My by default -->
            <add key="StoreName" value="My" />
            <!-- Optional. FindByThumbprint by default -->
            <add key="FindType" value="FindByThumbprint" />
            <!-- Required. -->
            <add key="FindValue" value="4894671ae5aa84840cc1079e89e82d426bc24ec6" />
        </fromStorage>
        <fromFile  name="Some key from file">
            <!-- Absolute or relative path to certificate file. -->
            <add key="Path" value=".\certificate.pfx" />
            <!-- Encrypted password -->
            <add key="Password" value="..." />
        </fromFile>
        <fromPEM  name="Some key from PEM">
-----BEGIN CERTIFICATE-----
MIIEjjCCA3agAwIBAgIJIBkBGf8AAAAPMA0GCSqGSIb3DQEBCwUAMIGzMQswCQYD
...
tJl1UvF7GWJd0yNyPVqCCnBY
-----END CERTIFICATE-----
        </fromPEM>
    </clientCertificates>
...
</configuration>
```

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Implementation must be reviewed prior test adding

BlackGad committed:
Implemented fromStorage and fromCert client certificates

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/signed)](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=3098) <br/>All CLA requirements met.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/ClientCertificate/ClientCertificateProvider.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public static class ClientCertificateProvider
+    {
+        #region Static members
```
We don't do too many region declarations in NuGet so I'd remove this.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -149,6 +149,12 @@
     <value>Encryption is not supported on non-Windows platforms.</value>
   </data>
+  <data name="Error_FromStorageCertificateNotFound" xml:space="preserve">
+    <value>Certificate was not found in {0}.{1} storage by {2} criteria with {3} value</value>
```
format: Space after a `.`.

```suggestion
    <value>Certificate was not found in {0}. {1} storage by {2} criteria with {3} value</value>
```

BlackGad replied:
Done

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -149,6 +149,12 @@
+  <data name="Error_FromStorageCertificateNotFound" xml:space="preserve">
+    <value>Certificate was not found in {0}.{1} storage by {2} criteria with {3} value</value>
+  </data>
+  <data name="Error_FromStorageSeveralCertificatesFound" xml:space="preserve">
```
Same thing applies.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/ClientCertificates.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <param name="httpClientHandler">Http client handler</param>
+        public static void SetupClientHandler(HttpClientHandler httpClientHandler)
+        {
+            if (httpClientHandler == null) throw new ArgumentNullException(nameof(httpClientHandler));
```
Use braces and new lines.

BlackGad replied:
Done

BlackGad replied:
Hm a bit confusing :) What do you mean?

nkolev92 replied:
```suggestion
            if (httpClientHandler == null)
            {
                throw new ArgumentNullException(nameof(httpClientHandler));
            }
```

BlackGad replied:
Done

nkolev92 reviewed (commented):
@BlackGad This looks amazing. 

I skimmed through it and it looks great so far. 
We probably should consider having a way to associate the source with the cert. 

Would be willing to write-up a spec similar to https://github.com/NuGet/Home/blob/dev/designs/NuGet-FrameworkReference.md, with some extra information and the scenarios? 

fyi @rrelyea @karann-msft @anangaur @chgill-MSFT

zivkan commented:
+1 to nkolev92's comment that it's amazing. Thanks for taking the time to create this. I agree that we should have a tech spec to look over first, to get the design agreed upon before worrying too much about the code.

Once the design is done, there are numerous things that need improvement before I'd be comfortable approving for merging, but this can be discussed at that time.

BlackGad committed:
Regions were removed from new code

BlackGad committed:
Added missed dots to error messages

BlackGad committed:
Added missed spaces after dots in error messages

BlackGad commented:
@nkolev92 
About 
> We probably should consider having a way to associate the source with the cert.

From my perspective it is redundant functionality that will add additional complexity. 

As far as I understand [HttpClientHandler](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler?view=netframework-4.8) will automatically manage when to use client certificate from [collection](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.clientcertificates?view=netframework-4.8). Extra certificates in this collection have no side effects for other servers. 

Also, based on my experience all clients will use client certificates in 2 ways: 
1. Per project nuget configuration (nuget.config under subversion)
2. Per user nuget configuration (%appdata%\NuGet\NuGet.Config)

BlackGad committed:
Fixed if/throw single line codestyle

BlackGad commented:
> Would be willing to write-up a spec similar to https://github.com/NuGet/Home/blob/dev/designs/NuGet-FrameworkReference.md, with some extra information and the scenarios?

Do you have design document requirements? (template, specs, etc)
Also what did you mean under additional scenarios? :)

nkolev92 commented:
@BlackGad Feel free to follow the template there. Just basic

Issue, Background, Requirements, Solution template is fine. 

The solution part is a bit free form based on the type of change. 

For example this change would require a CLI change to edit the config etc. So you'd have a part dedicated to that in the solution. Just something to get us going. 

By additional scenarios, I meant just call out the different scenarios for including the client certs, from store, from path etc. 

You've already done most of that in the PR description :) 

Also note that the same PR needs to work in dotnet.exe/nuget.exe and Visual Studio :)

BlackGad committed:
Added client certificates store in VSSolution

BlackGad commented:
Added VS solution client certificates from settings processing. 

As far as I understand same technique must be applied in:
src\NuGet.Core\NuGet.Build.Tasks\RestoreTask.cs - (use restoreContext.MachineWideSettings)
src\NuGet.Core\NuGet.Commands\CommandRunners\PushRunner.cs - (use settings)
src\NuGet.Core\NuGet.CommandLine.XPlat\Commands\PackageReferenceCommands\ListPackage\ListPackageCommand.cs - (after ProcessConfigFile)

But I can be wrong :) so domain developer will implement this in right way.

zivkan commented:
My concern is that the NuGet client SDk is also used by VS for Mac, Rider and others to add NuGet functionality into their apps. They have to know about this change, and do the initialisation somewhere in their code.

Is there another way to implement this so it doesn't require every process to add this additional line of code?

nkolev92 commented:
To add to Andy's point associating it with a source allows it to be carried by the PackageSource model, which would mean every client gets it for free.

BlackGad commented:
Well you need somehow link two operations: 
1) Load proper ISettings
2) Apply extracted information to HttpClientHandler.ClientCertificates in src\NuGet.Core\NuGet.Protocol\HttpSource\HttpHandlerResourceV3Provider.cs (hope they are using same codebase)

Every app loads settings in their way. 
My implementation similar to src\NuGet.Core\NuGet.Protocol\UserAgent.cs this.

Another way is assign ClientCertificates to src\NuGet.Core\NuGet.Configuration\PackageSource\PackageSource.cs. BUT it still require manual collection fill somewhere outside in client apps.

BlackGad commented:
Theoretically this singletone patch 
```CSharp
ClientCertificates.Store(ClientCertificateProvider.Provide(settings))
```
 can be placed inside 
```CSharp
PackageSourceProvider.ctor
```
and
```CSharp
PackageSourceProvider.LoadPackageSources()
```
Is there any other way how nuget clients can get PackageSource?

BlackGad commented:
Anyway will write spec later this week. 

Have no deep knowledge about NuGet architecture and external usage so seems it is all that I can do for now. Hope my work will help finally implement this feature.

BlackGad commented:
Specification [pull request](https://github.com/NuGet/Home/pull/8823).

BlackGad committed:
Accordingly to review fromCert was renamed to fromFile
fromCert PEM functionality was moved to new fromPEM item

BlackGad committed:
Added client-certificates add|remove|list command for command line client certificates management
Added required name attributes to all ClientCertificateItem's
Added ability to store plain password inside fromFile and fromPEM items

BlackGad committed:
Added ClientCertificatesCommandUsageExamples locales

BlackGad committed:
Added missed nuget command parameters description to resources

BlackGad commented:
Well... As @zivkan mention:

>Pretty much everything in the nuget.config file can be edited with nuget.exe commands, for example nuget config -set .., nuget sources add ... This spec should propose commands to add, list, edit and delete client certificate config.

Reasonable :) So I've implemented new command `client-certificates` with Add/Remove/List actions. Also added ability to specify passwords to `fromFile` and `fromPEM` items as plain text.

Also were added required `name` attributes to certificate source elements for proper identification via Add/Remove/List

Will change spec to describe all this stuff.

BlackGad committed:
Certificates on list now will return its thumbprints

BlackGad commented:
Locales partially implemented with Google translate, partially with same english description. New command API require native speakers translation.

dominoFire commented on src/NuGet.Core/NuGet.Commands/ClientCertificatesCommand/ClientCertificatesCommandRunner.cs:
```diff
@@ -0,0 +1,315 @@
+using System;
```
Missing License header

zivkan commented:
Is this PR superceeded by #3273 ?

BlackGad commented:
> Is this PR superceeded by #3273 ?

Yes it is. It was easier to write code from scratch.
